acl +fold

nigoroll opened the pull request:
Adds an optional feature to optimize ACLs by removing, for positive matches only, subnets already included in other supernets and merging adjacent entries.
See the second commit's message for detail.
Predecessors to these patches have been in production use since late 2019. This version has been adjusted to recent changes in the VCC ACL code: We now keep the acl entries in a colo^Wrank balanced tree instead of a list, which actually helped the clarity of this patch's code.
Because the changes applied by the `vcc_acl_merge` feature are not obvious from the test case, I add here the VCC output from https://github.com/varnishcache/varnish-cache/commit/8f64b0da6a7bbdcef2d2a521a2bbede8a9492ffa#diff-4f98af9b0a5f60a191eac6e289d0e649f7b9c280f3c61cc8b23fbee0b669f98c 

The ACL deliberately is constructed to contain various variants of the supernet & adjacent cases. I will show here the ACL entries and the relevant VCC output:

### The second entry is a subnet of the first


```
                # bad notation (confusing)
                "1.2.3.0"/24;
                "1.2.3.64"/26;
```
```
ACL entry:
('<vcl.inline>' Line 9 Pos 17)
                "1.2.3.0"/24;
----------------#########----

supersedes / removes:
('<vcl.inline>' Line 10 Pos 17)
                "1.2.3.64"/26;
----------------##########----
```

### Entry inserted as is, becomes relevant later

```
                "1.4.4.0"/22;
```

### Another subnet case

```
                "1.3.4.0"/23;
                "1.3.5.0"/26;
```
```
ACL entry:
('<vcl.inline>' Line 14 Pos 17)
                "1.3.4.0"/23;
----------------#########----

supersedes / removes:
('<vcl.inline>' Line 15 Pos 17)
                "1.3.5.0"/26;
----------------#########----
```

### Adjacent networks, get merged into `1.3.6.0/24`:

```
                "1.3.6.0"/25;
                "1.3.6.128"/25;
```
```
ACL entry:
('<vcl.inline>' Line 16 Pos 17)
                "1.3.6.0"/25;
----------------#########----

left of:
('<vcl.inline>' Line 17 Pos 17)
                "1.3.6.128"/25;
----------------###########----

removing the latter and expanding mask of the former by one to /24
```

### Supernet of two previous entries

```
                "1.3.0.0"/21;
```

Addition of this network removes `1.3.4.0/23` from above and `1.3.6.0/24` which we just merged. Note that the VCC reference is still to the source line, even though we changed the netmask. I think this is most POLA, but I am open to naming the changed entry, alternatively:

```
ACL entry:
('<vcl.inline>' Line 18 Pos 17)
                "1.3.0.0"/21;
----------------#########----

supersedes / removes:
('<vcl.inline>' Line 16 Pos 17)
                "1.3.6.0"/25;
----------------#########----

(That was just a warning)
ACL entry:
('<vcl.inline>' Line 18 Pos 17)
                "1.3.0.0"/21;
----------------#########----

supersedes / removes:
('<vcl.inline>' Line 14 Pos 17)
                "1.3.4.0"/23;
----------------#########----

(That was just a warning)
```

If you wonder about the order of the messages, this is because more specific ACL entries are found to the left of least specific ones and we walk right to left from the newly inserted entry.

### New entry superseded by previously added one

The next Two ACL entries tests that we properly remove new entries by one inserted much earier:

```
                "1.4.7";
                "1.4.6.0"/24;
```
```
ACL entry:
('<vcl.inline>' Line 13 Pos 17)
                "1.4.4.0"/22;
----------------#########----

supersedes / removes:
('<vcl.inline>' Line 19 Pos 17)
                "1.4.7";
----------------#######-

(That was just a warning)
ACL entry:
('<vcl.inline>' Line 13 Pos 17)
                "1.4.4.0"/22;
----------------#########----

supersedes / removes:
('<vcl.inline>' Line 20 Pos 17)
                "1.4.6.0"/24;
----------------#########----
```

### Merges of adjaent entries

The final four ACL entries test that we merge adjacent entries as given, but also the result of merges:
```
                # right,left adjacent
                "2.3.2.0"/23;
                "2.3.0.0"/23;
```
```
ACL entry:
('<vcl.inline>' Line 24 Pos 17)
                "2.3.0.0"/23;
----------------#########----

left of:
('<vcl.inline>' Line 23 Pos 17)
                "2.3.2.0"/23;
----------------#########----

removing the latter and expanding mask of the former by one to /22
```

```
                # left,right adjacent
                "2.3.4.0"/23;
                "2.3.6.0"/23;
```
```
ACL entry:
('<vcl.inline>' Line 26 Pos 17)
                "2.3.4.0"/23;
----------------#########----

left of:
('<vcl.inline>' Line 27 Pos 17)
                "2.3.6.0"/23;
----------------#########----

removing the latter and expanding mask of the former by one to /22
```

At this point we have `2.3.0.0/22` and `2.3.4.0/22`. Now these two ge merged into `2.3.0.0/21`:

```
ACL entry:
('<vcl.inline>' Line 24 Pos 17)
                "2.3.0.0"/23;
----------------#########----

left of:
('<vcl.inline>' Line 26 Pos 17)
                "2.3.4.0"/23;
----------------#########----

removing the latter and expanding mask of the former by one to /21
```

Again, the original ACL lines are quotes, despite having been changed to `/22`. Let me know if you think giving the changed entries would be better.

bsdphk commented:
Until now we have not done this, under the assumption that users wanted the VSL records to precisely report the entries in the acl.

bsdphk commented:
I think "prune" is a better name than "merge" ?

Given that ACL's no longer VSL log by default, it is not obvious to me if this optimization is worth doing for performance reasons.

Any performance gain must outweigh the risk of getting things wrong, and the effort to write tests that this is not the case.

nigoroll commented:
I do not care much about prune/merge. Technically it is actually both: For adjacent networks, It is a merge, for subnets, prune.
I need to report some numbers.

Dridi commented:
How about "fold" instead of merge or purge?

nigoroll commented:
I like _fold_

bsdphk commented:
I'm very paranoid about our ACL's being correct, so I did a couple of manual tests, and there seems to be an "off-by-8" kind of error:

        "10.0.0.0"/23;
        "10.0.2.0"/23;

        "10.1.0.0"/24;
        "10.1.1.0"/24;

        "10.2.0.0"/25;
        "10.2.0.128"/25;

Gives only:

    Message from VCC-compiler:
    ACL entry:
    ('/home/phk/Varnish/trunk/varnish-cache/bin/varnishd/a.vcl' Line 6 Pos 9)
            "10.0.0.0"/23;
    --------##########----

    left of:
    ('/home/phk/Varnish/trunk/varnish-cache/bin/varnishd/a.vcl' Line 7 Pos 9)
            "10.0.2.0"/23;
    --------##########----

    removing the latter and expanding mask of the former by one to /22
    (That was just a warning)
    ACL entry:
    ('/home/phk/Varnish/trunk/varnish-cache/bin/varnishd/a.vcl' Line 12 Pos 9)
            "10.2.0.0"/25;
    --------##########----

    left of:
    ('/home/phk/Varnish/trunk/varnish-cache/bin/varnishd/a.vcl' Line 13 Pos 9)
            "10.2.0.128"/25;
    --------############----

    removing the latter and expanding mask of the former by one to /24

nigoroll commented:
@bsdphk excellent find! I did not sufficiently test this case. :(
I would hope that this did not render the ACL implementation wrong, but certainly a merge opportunity was missed.
I have pushed a fix, but I am not yet sure if I am happy with it, so I will switch this PR to Draft for now.

nigoroll commented:
@bsdphk I found a _much_ simpler (2-line) fix, which I am happy with. I am leaving it as an extra commit for you to review easily, to be squashed before merge.

nigoroll commented:
> @bsdphk I found a _much_ simpler (2-line) fix, which I am happy with. I am leaving it as an extra commit for you to review easily, to be squashed before merge.

https://github.com/varnishcache/varnish-cache/commit/cdb5ba536290f33ab72aa69f417636fb370d4704 has been squashed now

nigoroll commented:
Friendly ping to @bsdphk

nigoroll commented:
I returned to this old PR after a *long* time to hopefully finally get it off the list and found a bug which was hidden from my production test by [the missed optimization found by phk](https://github.com/varnishcache/varnish-cache/pull/3563#issuecomment-982447991). The bug was that the *left of* relation did not consider the fact that the left network always needs to be even to qualify for a mask expansion. The fix is shown specifically in a17140ec59cb78e6f883ce36fc3ad160fdc6931a

I have also renamed the feature to `fold` as shown in 31c087a7a322fdfd8224a1b6358c936d60d95b1d

In the latest force-push d2929fad63cb8e6cafb0f1c78c27a77933f0deb0
I have squashed the above, rebased onto d99f62a9c5d4c6de492abae63510555d51a53324 and polished the commit messages.

This commit is going to production now.

nigoroll commented on doc/sphinx/reference/vcl.rst:
```diff
@@ -299,6 +299,27 @@ individually:
   However, if the name resolves to both IPv4 and IPv6 you will still
   get an error.
 
+* `+fold` - Fold ACL supernets and adjacent networks.
```
needs to be ``

nigoroll commented on bin/varnishtest/tests/c00005.vtc:
```diff
@@ -162,3 +162,170 @@ varnish v1 -errvcl {Non-zero bits in masked part} {
+		set req.http.ip = "10.0.3.255";	call t;
+		set req.http.ip = "10.1.1.255";	call t;
+		set req.http.ip = "10.2.0.255";	call t;
+}
```
indentation error

nigoroll reviewed (commented):
@simonvik's finds from vdd

nigoroll commented:
vdd: polish, add ipv6 test, merge

nigoroll commented:
vdd: polish, add ipv6 test, merge
